pandas/_libs/lib.pyx

```diff
@@ -1365,6 +1395,89 @@ cdef class Seen:
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
 
+    # These functions get the currently seen value
+    cdef uint8_t get_bool_val(self) except *:
+        if self.curr_seen is CurrSeen.bool_:
+            return self.curr_seen_val.bool_
+        else:
+            raise ValueError("Cannot convert non-booolean's to boolean")
+    cdef uint64_t get_uint_val(self) except *:
+        if self.curr_seen is CurrSeen.uint_:
+            return self.curr_seen_val.uint_
+        elif self.curr_seen is CurrSeen.int_:
+            return self.curr_seen_val.int_
+        elif self.curr_seen is CurrSeen.bool_:
+            return self.curr_seen_val.bool_
+        else:
+            # float, complex, NaN can't be cast to uint
+            raise ValueError("Cannot convert float/complex/NaN to uint64")
+
+    cdef int64_t get_int_val(self) except *:
+        if self.curr_seen is CurrSeen.uint_:
+            return self.curr_seen_val.uint_
+        elif self.curr_seen is CurrSeen.int_:
+            return self.curr_seen_val.int_
+        elif self.curr_seen is CurrSeen.bool_:
+            return self.curr_seen_val.bool_
+        else:
+            # float, complex, NaN can't be cast to uint
+            raise ValueError("Cannot convert float/complex/NaN to int64")
+
+    cdef float64_t get_float_val(self) except *:
+        if self.curr_seen is CurrSeen.null_:
+            return NaN
+        if self.curr_seen is CurrSeen.uint_:
+            return self.curr_seen_val.uint_
+        elif self.curr_seen is CurrSeen.int_:
+            return self.curr_seen_val.int_
+        elif self.curr_seen is CurrSeen.bool_:
+            return self.curr_seen_val.bool_
+        elif self.curr_seen is CurrSeen.float_:
+            return self.curr_seen_val.float_
+        else:
+            # complex cannot be cast to float
+            raise ValueError("Cannot convert complex to float64")
+
+    cdef complex128_t get_complex_val(self):
+        if self.curr_seen is CurrSeen.null_:
```

Comment:
Shouldn't these just have one condition to match the function name? Was expecting `get_complex_val` to just do:

```
if self.curr_seen is CurrSeen.complex_:
    return self.curr_seen_val.complex_
else:
    raise ValueError(...)
```

Reply:
No, since we can upcast nan/bool/int/uint/float to complex128.

Reply:
@WillAyd 
Maybe renaming these to ``as_complex``, ``as_float``, etc. would make more sense?

(I should've been more specific in my last comment, but the case that these would be triggered by is when e.g. you have [1.0, 2, 3]. When the curr_val becomes 2, you'll want to get the value as a float(even though curr_val is an int), since the array can only be inferred as a float/complex array at this point)

Reply:
Ah ok that is helpful thanks. Yea I think your suggestion makes sense. Do we also need each as_ function to have its own set of if conditions or can they be structured in a way that they can generically be placed in any implementation, just leaving it to the function to short circuit appropriately? 

Reason I ask is that the current structure generates a hierarchy of "available types" but in a way I don't think is very obvious. Probably better to trade some precision for readability/maintenance

Reply:
> Ah ok that is helpful thanks. Yea I think your suggestion makes sense. Do we also need each as_ function to have its own set of if conditions or can they be structured in a way that they can generically be placed in any implementation, just leaving it to the function to short circuit appropriately?
> 
> Reason I ask is that the current structure generates a hierarchy of "available types" but in a way I don't think is very obvious. Probably better to trade some precision for readability/maintenance

Yeah, I'm not too much too much a fan of this as well (unfortunately I don't think the current behavior can be changed). 
I'm also not thinking of a way to simplify it on top of my head, since use of the union requires the series of if statements. 

This hopefully shouldn't cause too big of an issue in the future, since
1) The numpy dtypes are unlikely to change, so this code shouldn't change much
2) Our test coverage should be good enough for this.

I've also embedded some comments in each function listing the rejected types for each function.

Given this, would you be fine merging this in its current state?